src/browser/engine/protocol.ts

```diff
@@ -1,24 +1,77 @@
 import { app, protocol } from 'electron';
 import { readFileSync, createReadStream } from 'fs';
-import stream from 'stream';
+import { Transform, PassThrough } from 'stream';
 import { lookup } from 'mime-types';
-import { join } from 'path';
-import { parse } from 'url';
+import { join, dirname, normalize } from 'path';
+import { parse, URL } from 'url';
+// @ts-ignore no types
+import matchAll from 'string.prototype.matchall';
 
 import { Protocol } from '../../common';
 import { getExtensionById } from '../chrome-extension';
-import { protocolAsScheme } from '../../common/utils';
-
-import ECx from './api';
-
+import { protocolAsScheme, fromEntries } from '../../common/utils';
+
+/**
+ * Protocol.ts
+ *
+ * Everything related to `chrome-extension://` protocol goes here
+ *
+ * - Register protocol
+ * - Handle and serve extension web resources and background page
+ *
+**/
+
+// defaultContentSecurityPolicy match Chromium kDefaultContentSecurityPolicy
+// https://cs.chromium.org/chromium/src/extensions/common/manifest_handlers/csp_info.cc?l=31
 // tslint:disable-next-line: max-line-length
 const defaultContentSecurityPolicy = 'script-src \'self\' blob: filesystem: chrome-extension-resource:; object-src \'self\' blob: filesystem:;';
 
-protocol.registerStandardSchemes(
+(protocol as any).registerStandardSchemes(
   [protocolAsScheme(Protocol.Extension)],
   { secure: true }
 );
 
+// The protocol handler load file into Buffers
+// before transform them into a Stream (expected in callback).
+// Stream callback allow custom headers.
+//
+// Handlable resources are listed in the extension manifest.
+// Allowed HTML pages can include extension assets that don't need
+// to be whitelisted. Since we didn't know the origin of the requests,
+// each time we serve a response, we scan the content to whitelist
+// inner resources.
+//
+// ref: https://developer.chrome.com/extensions/manifest/web_accessible_resources
+const whitelistedResourcesFromProtocolServedFiles = new Set();
+
+// Regex to capture the src attribute value for scrips, stylesheets...
+const resourceLinkRegex = /src\s*=\s*['"](.+?)['"]/g;
+
+const captureWhitelistableResourcesFromProtocolServedFile = (pathname: string) =>
```

Comment:
Woo this is quite complex, but I did understand it. Few questions:

> Allowed HTML pages can include extension assets that don't need to be whitelisted.
Are you sure it is necessary to do that? I see no reference of such practice in the documentation?

How does Chromium handle that?

> Since we didn't know the origin of the requests
Just in Electron, or even Chromium does not know?

Reply:
> Are you sure it is necessary to do that? I see no reference of such practice in the documentation?

["A navigation **from a web origin** to an extension resource will be blocked unless the resource is listed as web accessible"](https://developer.chrome.com/extensions/manifest/web_accessible_resources). So background page, popups and iFrames with `chrome-extension://` protocol bypass check.

> How does Chromium handle that?

https://cs.chromium.org/chromium/src/extensions/browser/extension_protocols.cc?rcl=4292bebbd8388070fc8718bb54d793b6f36fe4a6&l=391

> Just in Electron, or even Chromium does not know?

We don't know the origin of the requests only in Electron: referrer is always blank (issue: https://github.com/electron/electron/issues/14747). Chromium know the request context. 

-- 

I added all answers in comments belong to the code

Reply:
But as discussed yesterday, we should drop the check for now.
There is 3 reasons: we whitelist installable extensions so we can trust them, we resolve protocol request in the extension folder only at the user filesystem level and this code block may not cover all cases.

Reply:
Detail: I don't think https://github.com/electron/electron/issues/14747 is the right issue for this. `referrer` is an HTTP header that can be overridden some situation and is the frame URL's domain in another situation. We are rather looking for a strong reference to the frame / webContents from which the request came.